Retry queue: Connect transaction container to domain forwarder

ogaca-dd opened the pull request:
### What does this PR do?

This PR is part of **persisting failed transactions on disk when in-mem buffer is full**.
See #6867 for more context.

With this PR, when the retry queue from the domain_forwarder is full, the transactions are now stored on disk.

### Additional Notes

This PR uses the code from https://github.com/DataDog/datadog-agent/pull/6886

**Reviewing commit by commit is strongly recommended.**

### Describe your test plan

Write there any instructions and details you may have to test your PR.

hush-hush committed:
Adding support for multiple values per dogstatsd datagram

Numerical metrics (ie: not 'set') in dogstatsd now support sending
multiple values per message. This allows client to reduce traffic usage
and enable client side aggregation for histogram and distribution.

bmermet committed:
Customize msgpack parsing

bmermet committed:
Make nil map deserialization retrocompatible

bmermet committed:
Update api to use the new format

bmermet committed:
Fix tests

bmermet committed:
Remove code generation headers

bmermet committed:
Reorder imports

bmermet committed:
msgpack generation with marshall

bmermet committed:
Fix linter

bmermet committed:
Add release note

bmermet committed:
Remove sublayer tag from aggregation

bmermet committed:
Use pointers for stats data

bmermet committed:
Move sublayers in groupedStats

bmermet committed:
Fix benchmark

bmermet committed:
Small improvement for string concatenation

hush-hush committed:
Bumping datadog-go (DogStatsD Go client) to 4.2.0

lebauce committed:
[runtime-security] Support reloading rules (#6731)

* Support reloading rules

* Do not use the event type in the key for pids and inode discarders

* Add fail-fast option to stop at first failed test

* Add a one entry BPF array with enabled events

* Properly close SIGHUP channel

* Use non blocking send for rule match in tests

* Fix type in benchmark rules

* Allow running benchmark along with functional tests

* Avoid mutex on handle event code path

gbbr committed:
pkg/trace/{agent,writer}: remove Out channel (#6870)

This change removes the Out channel. It makes the code easier to read
and follow, whereas previously it was hard to figure out where the Out
channel was read from.

safchain committed:
Fix dentry ino when accessing lower layer (#6868)

safchain committed:
Remove the wilcard suffix limitation (#6817)

lebauce committed:
Do not cache fake inodes (#6845)

kolyshkin committed:
vendor: bump mountinfo to v0.4.0 (#6495)

It has some breaking changes, so updated pkg/security/probe
accordingly.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

clamoriniere committed:
Add orchestrator explorer k8s manifests example (#6873)

L3n41c committed:
[e2e tests] Enable all agents and check for pod readiness (#6851)

safchain committed:
Pre-allocate dentry error (#6874)

olivielpeau committed:
Delete sender/checkSampler when check is unscheduled (#6877)

Otherwise, the sender and checkSampler of a check leak when the check
is unscheduled.

Since the checkSampler stores metrics (and tags), not doing this can
lead to a significant memory leak when checks sending lots of metrics
(especially with lots of tags) are frequently scheduled/unscheduled.

olivielpeau committed:
[omnibus] Upgrade libkrb5 dependency (#6862)

To 1.8.3.

+ remove patch made redundant by the upgrade, since a
similar patch was merged upstream in https://github.com/krb5/krb5/commit/b43ebf0f0adebab7df02ada88bf639ddd67e23e0

clamoriniere committed:
Update docker dogstatsd based image (#6875)

juliogreff committed:
Break down tagger telemetry per source and prefix

We suspect that there is an entity leak in the tagger, where some
entities never get deleted, even though the actual entities (as in,
pods, containers, tasks, etc) are long gone. To help in the
investigation, we want to find out which collector and which kind of
entity is leaking, so we're adding extra telemetry to point us in the
right direction.

lebauce committed:
Fix parent filename discarder as regular expression (#6885)

p-lambert committed:
[networks] Add kitchen tests for system-probe (#6814)

* [networks] Add kitchen tests for system-probe

* [networks] Integrate system-probe kitchen tests into CI

* [networks] Upgrade default vagrant boxes for system-probe tests

* [networks] Add CI jobs for major platforms

* [networks] Run linter

* [networks] Fix DNS tests on Ubuntu 16.04

* [networks] Cleanup kitchen artifacts

* [networks] Fix conntrack integration test setup

* [networks] Copy build files

* [networks] Fix `TestDNSOverUDPTimeoutCount`

* [networks] Enable IPv6 support in kitchen tests

* [networks] Avoid using `dummy0` interface

* [networks] Refactor `TestUDPSendAndReceive` setup script

* [networks] Use `socat` in `setup_netns.sh`

* [networks] Auto enable tests for PRs tagged with `networks`

Commit:
[ecs] Expire cache key for availability of ECS EC2 resource tags endpoint (#6865)

gbbr committed:
pkg/trace/api: enable Windows pipe support (#6615)

Windows pipes are now supported by means of:

* `DD_APM_WINDOWS_PIPE_NAME` defines the pipe name to be used.
* `DD_APM_WINDOWS_PIPE_BUFFER_SIZE` defines the input buffer size
  (defaults to 1MB).
* `DD_APM_WINDOWS_PIPE_SECURITY_DESCRIPTOR` allows setting the security ID. Defaults to `D:AI(A;;GA;;;WD)` (grants access to everyone)

juliogreff committed:
Only update telemetry if source was actually deleted

The tag store previously used to always decrease the number of stored
entities even when sourceTags for a source had already been deleted
previously, causing telemetry to slowly decrease to over time to below
zero, which should be impossible. Now, telemetry is only updated if
there actually was a source to be deleted in the first place.

ahmed-mez committed:
[KSMCore] Metrics documentation (#6884)

* [KSMCore] Metrics documentation

* configure init config

remeh committed:
[hostname] adding a comment on the warning of the future usage of the fqdn (#6892)

Gui774ume committed:
[runtime-security] Expose process execution events and introduce process ancestors metadata (#6796)

* introduce ancestors cache

* add container support for process lineage

* add recursive delete and inherit container ID

* add procfs fallback

do not dispatch fork events

* add process lineage test on fork, exec and exit events

* fix proc fs excessive fall back

* fix touch resolution bug

* use newTestModule instead of newTestProbe

* introduce EVENT_MAX_ROUNDED_UP for pid discarders

* review changes

prognant committed:
[jmxfetch] Fix identical config file related crash (#6834)

ogaca-dd committed:
Update pipe name env (#6830)

* Rename dogstatsd_windows_pipe_name to dogstatsd_pipe_name.

urosgruber committed:
Fix FreeBSD build (#6787)

* Fix rtloader build
* Implement file_handles check
* Tweak default paths

Co-authored-by: Pablo Baeyens <pbaeyens31+github@gmail.com>

brycekahle committed:
Add manual stage to build ebpf-only clang 11 (#6888)

djova committed:
rtloader: preserve LC_RPATH for `datadog-agent-three` install on mac (#5738)

Set `INSTALL_RPATH_USE_LINK_PATH TRUE` for building `datadog-agent-three` macos to fix "image not found" error. Added it to `datadog-agent-two` as well for consistency even though on my local environment python2 was working OK without that flag.

I started getting this error when trying to run python checks locally:
```
Error: Could not initialize Python: could not load runtime python for version 3: Unable to open three library: dlopen(libdatadog-agent-three.dylib, 9): Library not loaded: @rpath/libpython3.8.dylib
  Referenced from: /Users/dusan.jovanovic/go/src/github.com/DataDog/datadog-agent/dev/lib/libdatadog-agent-three.dylib
  Reason: image not found
```

It was all working fine and somehow when I tried to build specifically for python2 then this problem started happening. I don't understand why cmake sometimes refuses to preserve the runtime search path. On my laptop it will works correctly for python2 but never python3. On others' it works correctly for both.

Currently on my laptop for python3, LC_RPATH is stripped away on install so it fails to load the python lib. Setting `INSTALL_RPATH_USE_LINK_PATH TRUE` preserves the LC_RPATH so the python3 lib loads successfully.

```bash
otool -L ./rtloader/build/three/libdatadog-agent-three.dylib | grep libpython
	@rpath/libpython3.8.dylib (compatibility version 3.8.0, current version 3.8.0)

otool -l ./rtloader/build/three/libdatadog-agent-three.dylib | grep LC_RPATH -A2
          cmd LC_RPATH
      cmdsize 56
         path /usr/local/anaconda3/envs/agent38/lib (offset 12)

otool -l ./dev/lib/libdatadog-agent-three.dylib | grep LC_RPATH -A2

otool -l ./dev/lib/libdatadog-agent-three.dylib | grep LC_RPATH -A2
          cmd LC_RPATH
      cmdsize 56
         path /usr/local/anaconda3/envs/agent38/lib (offset 12)
```

L3n41c committed:
Address the e2e tests failures of today (#6893)

* Force systemd to not print utf-8 characters: it makes GitLab crash!
* Ensure `curl` times out in all circumstances

juliogreff committed:
Don't double convert tagger IDs

At this point in the collector, we're no longer dealing with raw IDs (as
in, the ones containing the actual container runtime instead of just
`container_id`), but rather already the ones converted for use in the
tagger, so that double conversion was superfluous and was actually
making the tagger leak entities since it would try to delete entities it
couldn't find.

lebauce committed:
[runtime-security] Stop using the linux_bpf build tag (#6891)

* Stop using linux_bpf build tag

* Add missing comments

* golanci-lint fixes

* Use docker cookbook < 7.3.0

* Merge _bpf.go and non _bpf.go files

gbbr committed:
pkg/trace/agent: fix flaky test (#6902)

hmahmood committed:
[system-probe] Fix kitchen tests on debian 10 (#6899)

p-lambert committed:
[networks] Change system-probe CI test rules (#6895)

mgarabed committed:
Add option to output check results in a tabular format  (#6836)

* Add table formatter output

* Update imports

* Update licenses

* Update via go mod tidy

* Export the function correctly

* Fix linting

* Sort results and simplify table

* Enhance sorting

* Add release notes

* Update releasenotes/notes/tabular-check-format-4177ef9c7f8d0506.yaml

Co-authored-by: Ofek Lev <ofekmeister@gmail.com>

* Address comments

* Use strconv instead of Sprint

* Update metric sorting

* Format service check output

* Add blank lines to space output

Co-authored-by: Ofek Lev <ofekmeister@gmail.com>

derekwbrown committed:
Retain the domain/user that was used to register the agent service on (#6769)

* Retain the domain/user that was used to register the agent service on
initial install; this way, caller doesn't need to know it to upgrade

* Review feedback

* refactor into simpler functions

remeh committed:
[backport] CHANGELOG and release.json from `7.24.0` into `master`. (#6900)

* [backport] CHANGELOG and release.json from `7.24.0` into `master`.

* CHANGELOG: release date.

KSerrania committed:
[circleci] Update XCode version (#6908)

Updates the XCode version used in the CI to 11.7.0, which updates the underlying MacOS image (from 10.13 to 10.15).

ogaca-dd committed:
Make transaction storage optional

ogaca-dd committed:
Create `extractTransaction` from `flushToStorage`

ogaca-dd committed:
If disk serialization failed or is disabled, make sure currentMemSizeInBytes < maxMemSizeInBytes.

bmermet committed:
Make the bufferpool global

ogaca-dd committed:
Simplify transactionContainer (Behavior is the same)

ogaca-dd committed:
In transactionsFileStorage, remove the file from `filenames` even if `Deserialize` failed.

bmermet committed:
Move functions to decoder_bytes

bmermet committed:
Wrap buffer creation in newBuffer

bmermet committed:
Add comment to span_gen.go

ogaca-dd committed:
Instanciate `transactionContainer`

bmermet committed:
Add test to validate the custom msgpack implementation

bmermet committed:
Adapt msgpack tests

brycekahle committed:
Re-organize code to have `pkg/ebpf` only be common eBPF code (#6837)

bmermet committed:
Fix rebase problem

bmermet committed:
Wrap bufferpool accesses in functions

pierreguilmin committed:
Monitor dogstatsd buffer usage (#6438)

* Monitor dogstatsd buffer usage

- Fix documentation mistake on dogstatsd memory usage
- Add a metric to count the number of objects in the packets channel `packets_channel size`
- Add two metrics to monitor the packets buffer flush events: number of flush triggered by the timer (`packets_buffer_flush_timer`) vs number of flush triggered because the buffer is full (`packets_buffer_flush_full`)

* Add PacketsBuffer flush benchmark

ogaca-dd committed:
Use transactionContainer inside domain_forwarder.

hush-hush committed:
Fixing linter issue about int to string conversion

ogaca-dd committed:
Handle the transaction priority in transaction_container

ogaca-dd committed:
Add changelog

bmermet committed:
Fix test

bmermet committed:
Add copyright header

bmermet committed:
Address style comments

hmahmood committed:
[system-probe] Read udp ports bindings for all namespaces on init (#6764)

ogaca-dd committed:
Do not create `transactions_to_retry` folder if storage on disk is disabled.

ogaca-dd committed:
Fix an issue with nil interface

bmermet committed:
pkg/trace: optimize ExtractSubtraces (#6823)

* pkg/trace: reduce number of heap allocations in subtrace extraction

* Special case for single spans

* Fix test

bmermet committed:
Add release note

apigirl commented on releasenotes/notes/add_forwarder_storage_max_size_in_bytes-d860cf200534df6b.yaml:
```diff
@@ -0,0 +1,12 @@
+# Each section from every releasenote are combined when the
```
```suggestion
# Each section from every release note are combined when the
```

apigirl reviewed (commented):
just one small docs feedback, otherwise lgtm

prognant committed:
Revert "[logs/k8s] Retry query to kubelet (#6626)" (#6929)

AlexandreYang committed:
Delete systemd binary test file (#6928)

bmermet committed:
Address comments

ogaca-dd committed:
Update releasenotes/notes/add_forwarder_storage_max_size_in_bytes-d860cf200534df6b.yaml

Co-authored-by: maxime mouial <hush-hush@users.noreply.github.com>

ogaca-dd committed:
Update pkg/config/config_template.yaml

Co-authored-by: maxime mouial <hush-hush@users.noreply.github.com>

djova committed:
new SQL Exec Plan obfuscation API for checks (#5692)

Adds a new python API `datadog_agent.obfuscate_sql_exec_plan` to enable checks to obfuscate & normalize JSON Query Execution plans.

KSerrania committed:
[gitlab] Update & fix builders (#6931)

Updates all builders to new versions. This includes https://github.com/DataDog/datadog-agent-buildimages/pull/88 to fix problems with the build_system-probe-arm64 job.

The builders for the libbcc job have been pinned to a previous version of the system-probe image. The current version doesn't work.

ogaca-dd committed:
Handle review comments.

ogaca-dd committed:
Rename `f` to `tc` in transaction_container.go

ogaca-dd committed:
Remove `retryQueueSize` from `tryAddToTransactionContainer`

shang-wang committed:
[procutil] add initial models, test files and command line test (#6618)

* add procutil model to datadog-agent repo and add some tests

* remove all untested info for now

* remove all irrelevant fields to match the tests

* nit

* lint errors

* only for linux

* reduce test procfs size

* rename file to process_linux.go and remove common.go file as it's
redundant

* add more tests

* nit

* remove debug

* more tests

* nit

* add testFS and localFS tests as well

* skip tests on LocalFS test

* feedback

* nit

gbbr committed:
pkg/trace/obfuscate: allow more context in top-level call SQL escape sequences (#6770)

Historically we've been fully obfuscating [SQL escape sequences](https://docs.oracle.com/cd/E13157_01/wlevs/docs30/jdbc_drivers/sqlescape.html). This can be problematic for some when their entire query is an escape sequence calling a procedure, such as:

    {[?=]call procedure-name[([parameter][,parameter]...)]}.

In such cases, the query would amount to `?` which is not very insightful and compacts many queries under the same ambiguous resource.

This change adds an exception such that when a query starts with a curly brace, it is not fully obfuscated as an escape sequence, but instead is treated like a regular query where only strings and numerics inside of it get obfuscated.

[Various databases and languages](http://media.datadirect.com/download/docs/slnk/devref/scalarfn.html) support these types of calls.

gbbr committed:
pkg/trace/test: ensure that each run installs a fresh agent (#6906)

This change ensures that each integration test run installs a fresh
agent in a temporary directory as opposed to using or installing
whatever is found in the environments $GOPATH/bin/trace-agent

prognant committed:
[logs/AD] Logs tailing from docker label (#6883)

AlexandreYang committed:
Fix systemd check id (#6927)

albertvaka committed:
Add ddog-gov.com site option to Windows installer (#6936)

shang-wang committed:
[procutil] add commandline parsing benchmark for test fs and local fs (#6800)

* add procutil model to datadog-agent repo and add some tests

* remove all untested info for now

* remove all irrelevant fields to match the tests

* nit

* lint errors

* only for linux

* reduce test procfs size

* rename file to process_linux.go and remove common.go file as it's
redundant

* add more tests

* nit

* remove debug

* more tests

* nit

* add command line parsing bench for test fs and local fs

* add commandline parsing benchmark for test fs and local fs

* deal with errors for benchmark

* add testFS and localFS tests as well

* refined bench code

* add get PID benchmark

* skip tests on LocalFS test

* feedback

* nit

ahmed-mez committed:
[AD] Cleanup prometheus AD (#6937)

ogaca-dd committed:
Update releasenotes/notes/add_forwarder_storage_max_size_in_bytes-d860cf200534df6b.yaml

Co-authored-by: Kaylyn <kaylyn.sigler@datadoghq.com>

ogaca-dd committed:
Move the mutex from domainForwarder to transactionContainer.

ogaca-dd committed:
In FailedTransactionRemovalPolicy do not return on the first error

ogaca-dd committed:
Improve SerializeTo comment.

KSerrania committed:
[gitlab] Refactor config file (#6846)

First cleanup of the .gitlab-ci.yml file.

Make ordering of keys in job consistent:
- extends to describe the list of template jobs inherited
- allow_failure: true if the job is allowed to fail
- keys that specify when to run the job (rules, stage, resource_group)
- keys that specify where to run the job (image, tags)
- keys that specify job dependencies (needs, dependencies)
- keys about the job scripts (variables, before_script, script, after_script)
- keys about job result behavior (artifacts, retry, timeout)

Uniformize use of extends, remove <<: * when it can be replaced by extends
Remove NEW_BUILDER on Windows, it's not needed anymore
Remove useless allow_failure: false
Refactor docker_build_job_definition job definitions.
Move check_deploy stage to a more coherent place in the pipeline.

prognant committed:
[auditor] Allow to filter registry update by source ID (#6776)

p-lambert committed:
[networks] Log when DNS inspection is enabled but not supported (#6917)

L3n41c committed:
Add unit test for secret backend invocation from AD (#6914)

juliogreff committed:
Update storedTags on adds to every source (#6918)

Previously, because of the old telemetry that didn't have a breakdown
per source, storedTags was only updated when an entity didn't exist at
all in the store. When we started breaking down how many entities we
have stored per source, we'd actually just count the first source to
have seen the entity. This caused deletions, which did delete tags for
each source, to eventually bring the number of stored entities below
zero.

Now, each new source counts as an add, which will finally get us the
correct number of stored entities.

vboulineau committed:
Add trigger to build internal images (#6844)

ogaca-dd committed:
Move some instantiations logic from the forwarder to transactionContainer.

ogaca-dd committed:
Fix a typo in the release note (#6940)

Commit:
Fix the `security-agent runtime check-policies` command (#6932)

* Do not print usage on errors

* Do not print the error message on error

* Add a missing configuration parameter

* Use a small cache size

mfpierre committed:
Migrate orchestrator config to its own struct/package (#6910)

* Migrate orchestrator config to its own package

* Address feedback

* move extra tags to orchesrator config

safchain committed:
[runtime-security] use only one perf ring, clean dentry chain, cleanup dentry after unmount (#6948)

* Use only one perf ring for regular event and mount event to avoid race

* Invalidate all of the dentry of a mount point on umount

* Fix exec kprobe on bionic

* Delete recursively dentry cache

remeh commented on pkg/forwarder/domain_forwarder.go:
```diff
@@ -135,12 +147,39 @@ func (f *domainForwarder) retryTransactions(retryBefore time.Time) {
+		transactionEndpointName := t.GetEndpointName()
+		transactionsDroppedByEndpoint.Add(transactionEndpointName, int64(dropCount))
+		transactionsDropped.Add(int64(dropCount))
+		tlmTxDropped.Inc(f.domain, transactionEndpointName)
```
Shouldn't it be increased by dropCount instead of 1?

ogaca-dd replied:
Code updated.

AlexandreYang committed:
Add loader selection to init_config (#6700)

shang-wang committed:
[procutil] add parsing status code and tests (#6628)

* add procutil model to datadog-agent repo and add some tests

* remove all untested info for now

* remove all irrelevant fields to match the tests

* nit

* lint errors

* only for linux

* add parsing status code and tests

* nit and comments

* assert no errors

* reduce test procfs size

* fix dir

* rename file to process_linux.go and remove common.go file as it's
redundant

* add more tests

* nit

* fix merge conflict

* remove debug

* more tests

* nit

* add new test case

* add missing test case

* add testFS and localFS tests as well

* add local and test fs

* skip localFS test

* skip tests on LocalFS test

* feedback

* nit

* updated go.sum

* correct the parsing bug for NSpid, also add a test

* nit

* port NSpid test and test files from gopsutil

* add all fields to test case

ahmed-mez committed:
[AD][Tagger] Fix potential inconsistent tags hash (#6950)

ogaca-dd committed:
Make all methods private.

raphaelgavache committed:
Update top tag from tracer (#6941)

bmermet committed:
pkg/trace: switch v0.5 msgpack deserialization to read from []byte (#6946)

* pkg/trace: switch v0.5 msgpack deserialization to read from []byte

* Apply suggestions from code review

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

* Remove newlines

* Add release notes

Co-authored-by: Gabriel Aszalos <gabriel.aszalos@gmail.com>

ogaca-dd committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into olivierg/failed-transactions-connect-transactionContainer-to-domain-forwarder

ogaca-dd committed:
Remove the documentation for `forwarder_storage_max_size_in_bytes`

ogaca-dd committed:
update go.mod

ogaca-dd committed:
Fix unit tests on Windows.

ogaca-dd commented on pkg/forwarder/domain_forwarder.go:
```diff
@@ -36,36 +35,38 @@ type domainForwarder struct {
+	blockedList                  *blockedEndpoints
 }
 
-func newDomainForwarder(domain string, numberOfWorkers int, retryQueueLimit int, retryQueueAllPayloadsMaxSize int, connectionResetInterval time.Duration) *domainForwarder {
```
`newDomainForwarder` cannot provide a default implementation for `transactionContainer`. `nil` is a valid value for `transactionContainer`.

ogaca-dd commented on pkg/forwarder/forwarder.go:
```diff
@@ -229,15 +254,67 @@ func NewDefaultForwarder(options *Options) *DefaultForwarder {
+		}
+	}
+
+	flushToDiskMemRatio := config.Datadog.GetFloat64("forwarder_flush_to_disk_mem_ratio")
```
Code updated.

ogaca-dd commented on pkg/forwarder/domain_forwarder.go:
```diff
@@ -36,36 +35,38 @@ type domainForwarder struct {
 	retryQueue                   []Transaction
 	retryQueueLimit              int
 	retryQueueAllPayloadsMaxSize int
+	optionalTransactionContainer *transactionContainer
```
As discussed in private this code will be updated.

ogaca-dd commented:
Closed in favor of https://github.com/DataDog/datadog-agent/pull/6921.
